torch_xla/experimental/scan.py

```diff
@@ -82,14 +132,514 @@ def scan(fn, init, xs):
+  # forward and backward graphs. Note that the init/carry fake tensor
+  # always requires grad. That's because even if the user passed in some
+  # `init` that does not require grad, we still want gradients to flow
+  # through the `carry` from one iteration of the user function to the
+  # next. In summary, the `carry` argument used to trace a user function
+  # to get a correct backward pass always requires grad.
+  def make_fake_tensor(v: torch.Tensor, requires_grad=True) -> torch.Tensor:
+    return torch.empty_like(
+        v, dtype=v.dtype, device=v.device, requires_grad=requires_grad)
+
+  fake_carry_pytree = tree_map(make_fake_tensor, init)
+  fake_x_pytree = tree_map(
+      lambda v: make_fake_tensor(v[0], requires_grad=v.requires_grad), xs)
+
+  with torch.enable_grad():
+    fw_compiler, get_fwd = _make_get_graph_compiler()
+    bw_compiler, get_bwd = _make_get_graph_compiler()
+    fn_compiled = aot_function(
+        fn,
+        fw_compiler=fw_compiler,
+        bw_compiler=bw_compiler,
+        partition_fn=partition_fn)
+    _, bwd_out_spec = tree_flatten((fake_carry_pytree, fake_x_pytree))
+    out = fn_compiled(fake_carry_pytree, fake_x_pytree)
+    out, fwd_out_spec = tree_flatten(out)
+    torch.autograd.backward(out, tree_map(lambda v: torch.ones_like(v), out))
+    # How many outputs out of the fwd_graph is actually outputs of `fn`, and not
+    # intermediate activations.
+    num_out = len(out)
+
+  fwd_graph = get_fwd()
+  bwd_graph = get_bwd()
+
+  def forward(carry, x):
+    flat_carry, _ = tree_flatten(carry)
+    flat_x, _ = tree_flatten(x)
+    out = fwd_graph(*flat_carry, *flat_x)
+    actual_out, activations = split(out, num_out)
+    carry, y = tree_unflatten(actual_out, fwd_out_spec)
+    y = (y, activations)
+    return carry, y
+
+  def backward(carry, x):
+    grad_new_carry, _ = tree_flatten(carry)
+    (grad_y, activations) = x
+    grad_y, _ = tree_flatten(grad_y)
+    out = bwd_graph(*activations, *grad_new_carry, *grad_y)
+    grad_carry, grad_x = tree_unflatten(out, bwd_out_spec)
+    return grad_carry, grad_x
+
+  return forward, backward
+
+
+def _make_get_graph_compiler():
+  """
+  Creates a compiler that records the graph, and a getter
+  function to retrieve them.
+  """
+  graph: List[Optional[torch.fx.GraphModule]] = [None]
+
+  def forward_comp(fx_module: torch.fx.GraphModule, _):
+    assert graph[0] is None
+    graph[0] = fx_module
+    return make_boxed_func(fx_module)
+
+  def get_graph():
+    g = graph[0]
+    assert g is not None
+    return g
+
+  return forward_comp, get_graph
+
+
+@pytreeify
+class Scan(torch.autograd.Function):
+
+  @staticmethod
+  def forward(ctx, forward, backward, init, xs):
+    # Forward pass, save activations for backward
+    ctx._backward = backward
+    with torch.no_grad():
+      carry, ys = _scan_impl_pytree(forward, init, xs)
+      ys, activations = ys
+    ctx.save_for_backward(*activations)
```

Comment:
is this stacked `activations`?

Reply:
I am trying to figure out where the gradient checkpointing happens when you use a different partion function.